### PR TITLE
Implement ModelBuilder and update the execution API

### LINKED ISSUE #22: Implement the ModelBuilder and update the execution API
Implement the changes of https://github.com/webmachinelearning/webnn/pull/94.

### PR DESCRIPTION
**Pull request by huningxin**
fix #22 

Implement the WebNN API change introduced by https://github.com/webmachinelearning/webnn/pull/94. 

Preview:
* [test](https://huningxin.github.io/webnn-polyfill/test/) 
* [docs](https://huningxin.github.io/webnn-polyfill/docs/)

Detailed changes include:
1. Support optional outputs and optional input shapes.
2. Implement ModelBuilder interface for state separation between context state and graph state.
3. Fold the execution interface into the compilation interface by adding compute method.
4. Remove quantization-related params from the operand descriptor.
5. Simplify operand type enum.
6. Add 86 new test cases.
7. Update the version to 0.1.0

@anssiko @wchao1115 @pyu10055 @BruceDai , please take a look. Thanks!

### CODE REVIEW

**Comment by huningxin** [comment]
Thanks @BruceDai for contributing the nhwc conv2d and depthwise conv2d test cases. It helped me to find an issue of filter layout for nhwc depthwise conv2d. I pushed new comments to fix it. Please take a look.

**Review by anssiko — APPROVED** [review]
Thanks! I defer to other reviewers for a detailed look.

File: test/api/compilation.js [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+const expect = chai.expect;
+const assert = chai.assert;
+
+import * as utils from '../utils.js';
+
+describe('test Compilation', function() {
+  const nn = navigator.ml.getNeuralNetworkContext();
+  const builder = nn.createModelBuilder();
+  const desc = {type: 'float32', dimensions: [2, 2]};
+  const a = builder.input('a', desc);
```
Comment (pyu10055) [inline test/api/compilation.js:12]: operands seem to contain builder, can it provide chained API?
i.e. a.matmul(b)

File: test/api/compilation.js [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+  const d = builder.constant(
+      {type: 'float32', dimensions: [2, 2]}, new Float32Array(4).fill(1));
+  const e = builder.add(c, d);
+  const bufferA = new Float32Array(4).fill(1);
+  const bufferB = new Float32Array(4).fill(1);
+  const expectedC = [2, 2, 2, 2];
+  const expectedE = [3, 3, 3, 3];
+
+  it('Compilation should have compute method', async () => {
+    const model = builder.createModel({c});
```
Comment (pyu10055) [inline test/api/compilation.js:24]: this is bit confusing, the builder creates models, but the topology is created before the model is created, and how made the topology immutable for the model?

**Review by pyu10055 — COMMENTED** [review]
thanks for aligning with the API changes, I had a couple high level questions in the comments.

File: test/api/compilation.js (hunk 1 above)
Comment (huningxin) [reply test/api/compilation.js:12]: Created https://github.com/webmachinelearning/webnn/issues/106 to follow-up in spec.

File: test/api/compilation.js (hunk 2 above)
Comment (huningxin) [reply test/api/compilation.js:24]: The model is immutable after `createModel`. As discussed with @pyu10055 , it needs to add details of the `createModel` method into the spec. Created https://github.com/webmachinelearning/webnn/issues/107 to follow up.

Comment (huningxin) [reply test/api/compilation.js:24]: I added a new commit https://github.com/webmachinelearning/webnn-polyfill/pull/23/commits/e0ba924435282afff0bc3984218855a3bca947d1. It includes a test case for immutable model and fixes a bug that the constant operand is not immutable. Please take another look.

**Comment by huningxin** [comment]
@BruceDai , could you please verify whether the depthwise conv2d issue was fixed? Thanks!

**Comment by BruceDai** [comment]
@huningxin I verified with [three test cases](https://github.com/BruceDai/webnn-polyfill/blob/verify-depthwise_conv2d/test/android_cts/converted_conv2d.js) converted from [Android NNAPI CTS tests](https://android.googlesource.com/platform/frameworks/ml/+/refs/tags/android-cts-10.0_r5), now these test cases were all PASS, thanks for fixing this issue. 

BTW, I'm planning to submit a PR of adding some test cases converted from Android NNAPI CTS tests and the converter scripts to this repo, and hope to get some feedback from you and also the community, thanks.

**Review by BruceDai — APPROVED** [review]
Thanks @huningxin , this PR looks good to me.

**Comment by anssiko** [comment]
@huningxin I believe this PR has received adequate review and we can merge this now.

**Comment by huningxin** [comment]
@anssiko @BruceDai @pyu10055 , I believe all the comments have been addressed by either adding new commits or filing new issues. With that, I am going to merge this PR. Thanks much!

**Pull request merged**